lib/collections.c

```diff
@@ -129,3 +131,103 @@ int64_t conch_checkout(Conch* conch)
 bool conch_is_available(Conch* conch) { return conch->conch_available; }
 
 void conch_free(Conch* conch) { free(conch); }
+
+// Dynamic Array ADT
+typedef struct List
+{
+    size_t length;
+    size_t max_length;
+    void** data;
+} List;
+
+List* list_init(void)
+{
+    List* list = malloc(sizeof(List));
+    if (list == NULL)
+    {
+        fprintf(stderr, "Error: list_init: malloc failed\n");
+        exit(EXIT_FAILURE);
+    }
+    list->length = 0;
+    list->max_length = INITIAL_LIST_MAX_LENGTH;
+    list->data = malloc(INITIAL_LIST_MAX_LENGTH * sizeof(void*));
+    if (list->data == NULL)
+    {
+        fprintf(stderr, "Error: list_init: malloc failed\n");
+        exit(EXIT_FAILURE);
+    }
+    return list;
+}
+
+void* list_get_at(List* list, size_t index)
+{
+    if (index >= list->length)
+    {
+        fprintf(stderr, "Error: list_get_at: index out of bounds\n");
+        exit(EXIT_FAILURE);
+    }
+    return list->data[index];
+}
+
+void list_set_at(List* list, size_t index, void* item)
+{
+    if (index >= list->length)
+    {
+        fprintf(stderr, "Error: list_set_at: index out of bounds\n");
+        exit(EXIT_FAILURE);
+    }
+    list->data[index] = item;
+}
+
+static void list_resize(List* list)
+{
+    if (list->length * 2 >= list->max_length)
+    {
+        list->max_length *= 2;
+        list->data = realloc(list->data, list->max_length * sizeof(void*));
+        if (list->data == NULL)
+        {
+            fprintf(stderr, "Error: list_resize: realloc failed\n");
+            exit(EXIT_FAILURE);
+        }
+    }
+    else if (list->length * 4 <= list->max_length)
+    {
+        list->max_length /= 2;
+        list->data = realloc(list->data, list->max_length * sizeof(void*));
+        if (list->data == NULL)
+        {
+            fprintf(stderr, "Error: list_resize: realloc failed\n");
+            exit(EXIT_FAILURE);
+        }
+    }
+}
+
+void list_append(List* list, void* item)
+{
+    list->data[list->length++] = item;
+    list_resize(list);
+}
+
+void* list_pop(List* list)
+{
+    if (list->length == 0)
+    {
+        fprintf(stderr, "Error: list_pop: list is empty\n");
+        exit(EXIT_FAILURE);
+    }
+    return list->data[--list->length];
+    list_resize(list);
+}
+
+size_t list_length(List* list) { return list->length; }
+
+void list_free(List* list)
+{
+    for (size_t i = 0; i < list->length; i++)
+    {
+        free(list->data[i]);
+    }
```

Comment:
I'm not entirely sure if it's `list_free`'s responsibility to do this!

Reply:
Great question.

A similar thing comes up in `list_set_at`, since a (void*) pointer gets overwritten. If we wanted, we could call `free` there as well, right before writing the new (void*) value to that cell.

But I think for now, the simplest thing is to never call `free` on the user-data. So I'm in favour of removing this loop.